### DISCUSSION TITLE
Improve exception handling and type safety in python SDK

### DISCUSSION THREAD

**Discussion by hubert-springbok** [langfuse/langfuse#874] [Ideas]
# TL;DR
The error handling in the python SDK causes type errors and unreliable code. Let's fix it.

# The problem
The existing Langfuse python SDK implements a pattern which we found quite inconvenient: most of the public methods, e.g. `.span` or `.create_trace` are wrapped in a generic `try-catch` block and in case of any error, they log a message and return `None`.

The problem with this is that the burden of checking whether everything went fine is pushed back into application code.
Consider this snippet from the docs:
```
trace = langfuse.trace(name = "llm-feature")
retrieval = trace.span(name = "retrieval")
retrieval.generation(name = "query-creation")
```

It looks good on paper, but Langfuse SDK doesn't actually work like this atm. If anything ever goes wrong in `.trace`, the next time the application code tries to call `trace.span`, there will be a runtime error and the entire program will crash. The type checker does complain about this as well.

In reality, the above snippet must be written as:
```
trace = langfuse.trace(name = "llm-feature")
retrieval = trace.span(name = "retrieval") if trace else None
if retrieval:
    retrieval.generation(name = "query-creation")
```
to be any safe. Still, figuring out what went wrong and whether monitoring is still on is somewhat difficult.

# Improvement and backwards compatibility
I'd like to propose changes in the SDK implementation that will make the syntax from the docs reality. The condition for that is simple: the public API methods that are expected to return a Client, Trace, Span etc. must do so in all cases, even if there is an exception, or raise an exception themselves.

The easiest implementation would be of course to remove the try-catch blocks and raise, like most normal python libraries would do. However, in this case the change would be backwards-incompatible and I understand exceptions are still considered 'a bad thing' in some circles.

# Proposed solution
To maintain the current behaviour some users may expect, but provide type safety and correctness, I recommend the following implementation:
 * introduce a new configuration variable, e.g. `LANGFUSE_PROPAGATE_EXCEPTIONS`, with default to False.
 * if LANGFUSE_PROPAGATE_EXCEPTIONS is True, raise any unhandled exceptions in the SDK and let the application code deal with them. This will give full control to more advanced users.
 * if LANGFUSE_PROPAGATE_EXCEPTIONS is False and there is an exception in the SDK, log it and return a 'dummy' object with the same public API as the real thing.
 
 The dummy accepts all calls that would be valid with the real object and just does nothing. Of course, this will drop monitoring events for the section where the error occured, but that's already the behaviour of the SDK as it stands.

# Implementation
I'm happy to contribute to the implementation of the above proposal. We've internally created a DummyTaskManager already and given that all spans, generations etc. rely on that, it covers most of the proposed scope.

**Comment by marcklingen** [comment]
Thanks for the suggestion, this is great! I agree that the runtime behavior of the library should not depend on the background network calls to the Langfuse API.

**Comment by maxdeichmann** [comment]
Hi @hubert-springbok, thanks again for raising this. I pushed a change to fix this recently (https://github.com/langfuse/langfuse-python/releases/tag/v2.15.0). Feel free to check that out and tell me whether it worked for you :)

**Comment by hubert-springbok** [comment]
Thanks, that's much appreciated! We'll update next week and simplify our code a fair bit.
I've created a small [follow-up MR with some cleanup and explicit type hints](https://github.com/langfuse/langfuse-python/pull/405)
